### PR TITLE
fix(sync): treat empty .env.keys value as present; collision-safe backups

### PR DESCRIPTION
**Pull request by jainal09**
Closes two of the 5 remaining #413 reliability findings (15, 16), both in `sync/operations.py`.

## Finding 15 — empty value read back as `None`

`EnvKeysFile.read_key` used `^KEY=(.+)$`, which requires at least one character after `=`. A present-but-empty key (`KEY=`) therefore read back `None` — but `None` is the engine's signal for **key absent**. Consequences (engine.py):
- An empty vault secret re-synced **forever as `CREATED`** (write `KEY=`, re-read `None`, create again).
- `verify-only` reported a false `Key file does not exist` ERROR even though the key was present.

**Fix:** use `^KEY=(.*)$` so `KEY=` returns `""` (present), while a genuinely absent key still returns `None`. The engine then takes the `local_value == vault_value` path → `SKIPPED` when both are empty.

## Finding 16 — same-second backup overwrite

`create_backup` stamped backups with `%Y%m%d_%H%M%S` (second precision), so two backups within the same second resolved to the **same path** and `shutil.copy2` silently overwrote the first.

**Fix:** microsecond precision (`%f`) plus an incrementing suffix on the off-chance of a same-tick collision, so an earlier backup is never clobbered.

## Tests

`tests/unit/test_sync_operations.py`:
- `test_read_key_empty_value_returns_empty_string` — `KEY=` → `""`, absent key → `None`
- `test_write_then_read_empty_value_roundtrips`
- `test_create_backup_same_timestamp_no_overwrite` — frozen clock forces the collision; both backups exist with their own content (mocks time, not the backup logic)

`operations.py` 96% line coverage; broader sync suite (operations + engine + vault-push) green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed handling of empty environment variable values—now correctly preserved as empty strings instead of treated as missing.
  * Improved backup file creation with enhanced timestamp precision to prevent collisions when multiple backups are created simultaneously.
  * Failed backup operations now properly clean up temporary artifacts, preventing incomplete files from being left behind.

<h3>Greptile Summary</h3>

This PR fixes two reliability bugs in `sync/operations.py`: an empty-value sentinel mismatch in `read_key` that caused perpetual re-sync for empty vault secrets, and a same-second backup overwrite that has now been upgraded to microsecond precision with O_EXCL atomic claiming and counter-based suffix fallback.

- **Finding 15 (`read_key`):** Regex widened from `(.+)` to `(.*)` so `KEY=` returns `""` (present) rather than `None` (absent), preventing the false `CREATED` loop and spurious `Key file does not exist` errors. The surrounding `.strip()` on an already-empty group is a no-op and does not regress the fix.
- **Finding 16 (`create_backup`):** Timestamp upgraded to microsecond precision; the O_EXCL + counter loop (previously reviewed and hardened atomically) is exercised by three new deterministic tests that freeze the clock, inject copy failures, and assert no zero-byte placeholder survives an error.

<h3>Confidence Score: 5/5</h3>

Both changes are narrow, well-understood bug fixes with deterministic tests; no existing behaviour is regressed.

The regex change is a one-character delta (`+` → `*`) whose only effect is returning `""` instead of `None` for an empty value — the surrounding `.strip()` on an already-empty string is a no-op. The backup hardening (microsecond timestamp, O_EXCL atomic claim, counter suffix, placeholder cleanup on failure) was already reviewed and approved in earlier rounds; this PR adds the missing microsecond precision that closes the original same-second overwrite window.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/envdrift/sync/operations.py | Two targeted fixes: regex widened to `(.*)` for empty-value correctness, and backup stamping upgraded to microsecond precision with O_EXCL atomic claiming and counter collision fallback plus placeholder cleanup on copy failure. |
| tests/unit/test_sync_operations.py | Three new tests added: empty-value roundtrip, frozen-clock backup collision, and copy-failure placeholder cleanup — each deterministically exercises the exact failure mode described in the PR. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["read_key(key_name)"] --> B{File exists?}
    B -- No --> C["return None"]
    B -- Yes --> D["splitlines, match ^KEY=(.*)$"]
    D --> E{Line matched?}
    E -- No --> C
    E -- Yes --> F["group(1).strip()"]
    F --> G{Quoted?}
    G -- Yes --> H["strip quotes"]
    G -- No --> I["return value ('' if empty)"]
    H --> I

    J["create_backup()"] --> K{Source exists?}
    K -- No --> L["raise FileNotFoundError"]
    K -- Yes --> M["timestamp = strftime('%Y%m%d_%H%M%S_%f')"]
    M --> N["backup_path = name.backup.timestamp"]
    N --> O["os.open O_CREAT|O_EXCL|O_WRONLY"]
    O --> P{FileExistsError?}
    P -- Yes --> Q["backup_path = name.backup.timestamp.counter\ncounter += 1"]
    Q --> O
    P -- No --> R["os.close fd"]
    R --> S["shutil.copy2 source → backup_path"]
    S --> T{Copy succeeded?}
    T -- Yes --> U["return backup_path"]
    T -- No --> V["unlink backup_path placeholder\nraise"]
```

<sub>Reviews (4): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/sync-operat..."](https://github.com/jainal09/envdrift/commit/527b17ce33d6ba145e0c168eac35b511e4620df1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36296838)</sub>

### CODE REVIEW

File: src/envdrift/sync/operations.py [hunk 1]
```diff
@@ -108,12 +112,22 @@ def has_dotenvx_header(self) -> bool:
...
             raise FileNotFoundError(f"Cannot backup non-existent file: {self.path}")
 
-        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S_%f")
         backup_path = self.path.parent / f"{self.path.name}.backup.{timestamp}"
+        counter = 1
+        while backup_path.exists():
+            backup_path = self.path.parent / f"{self.path.name}.backup.{timestamp}.{counter}"
+            counter += 1
         shutil.copy2(self.path, backup_path)
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/sync/operations.py]: Fixed in b461291 — the backup target is now claimed atomically with os.open(O_CREAT | O_EXCL), falling through to the next suffix on FileExistsError, closing the existence-check/copy TOCTOU window.

File: src/envdrift/sync/operations.py [hunk 2]
```diff
@@ -108,12 +112,32 @@ def has_dotenvx_header(self) -> bool:
...
+                # TOCTOU window between an existence check and the copy.
+                fd = os.open(str(backup_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
+            except FileExistsError:
+                backup_path = self.path.parent / f"{self.path.name}.backup.{timestamp}.{counter}"
+                counter += 1
+                continue
+            os.close(fd)
+            break
         shutil.copy2(self.path, backup_path)
         return backup_path
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/sync/operations.py:149]: Fixed in bf9f577 — the copy is now wrapped so a failure unlinks the empty O_EXCL placeholder before re-raising (mirrors atomic_write). Regression test injects a copy failure and asserts no leftover backup file.

**Comment by jainal09** [comment]
🤖 **envdrift review-bot** — automated re-review of latest head `bf9f5774`


Reviewed the PR description and cited #413 context, the full diff, and the latest cleanup commit `bf9f577`. The previous zero-byte backup placeholder issue is fixed at this head, but I found one remaining sync correctness issue around the same “empty value is present” behavior.

Verification performed:

- `uv run --extra dev pytest tests/unit/test_sync_operations.py -q` → `37 passed`
- Re-ran the previously failing backup-copy repro by monkeypatching `shutil.copy2` to raise `OSError("disk full")`; `create_backup()` raised and left `backup_count=0`.

**Findings (1):**

- [medium] Full-line empty vault keys are still treated as non-empty prefixed strings.

`src/envdrift/sync/engine.py:280` strips a full-line vault value with `re.match(r"^DOTENV_PRIVATE_KEY_([A-Za-z0-9_]+)=(.+)$", value)`. Because the value side uses `(.+)`, `DOTENV_PRIVATE_KEY_PRODUCTION=` does not match and is returned as the literal vault value instead of the empty key material. That undermines this PR’s empty-value semantics for vault values that store the dotenvx full-line format.

Local repro at `bf9f5774`: with `.env.keys` containing `DOTENV_PRIVATE_KEY_PRODUCTION=`, vault value `""` produces `SyncAction.SKIPPED` with `<empty>` previews, but vault value `"DOTENV_PRIVATE_KEY_PRODUCTION="` produces `SyncAction.ERROR`, `Value mismatch detected`, and a redacted vault preview for the whole prefixed string. In force-update mode this can rewrite the local file to `DOTENV_PRIVATE_KEY_PRODUCTION=DOTENV_PRIVATE_KEY_PRODUCTION=`. Please allow an empty captured value, e.g. `(.*)`, and add a regression test for full-line empty vault values.

Current live checks: integration matrix and CI coverage were still in progress at the last poll.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>
